File: trunk/dotnet/picocontainer/src/Test/Defaults/SetterInjectionComponentAdapterFactoryTestCase.cs
using System;
using PicoContainer.Tests.Tck;
using PicoContainer.Defaults;
using PicoContainer;
using NUnit.Framework;

namespace Test.Defaults
{
	/// <summary>
	/// Summary description for SetterInjectionComponentAdapterFactoryTestCase.
	/// </summary>
	[TestFixture]
	public class SetterInjectionComponentAdapterFactoryTestCase : AbstractComponentAdapterFactoryTestCase
	{
		[SetUp]
		protected void setUp()
		{
			picoContainer = new DefaultPicoContainer(CreateComponentAdapterFactory());
		}

		protected override IComponentAdapterFactory CreateComponentAdapterFactory()
		{
			return new SetterInjectionComponentAdapterFactory(new ConstructorInjectionComponentAdapterFactory());
		}

		public interface Bean
		{
		}

		public class NamedBean : Bean
		{
			private String name;

			public String Name
			{
				get { return name; }
				set { name = value; }
			}
		}

		public class NamedBeanWithPossibleDefault : NamedBean
		{
			private bool byDefault = false;

			public NamedBeanWithPossibleDefault()
			{
			}

			public NamedBeanWithPossibleDefault(String name)
			{
				Name = name;
				byDefault = true;
			}

			public bool ByDefault
			{
				get { return byDefault; }
			}
		}

		public class NoBean : NamedBean
		{
			public NoBean(String name)
			{
				Name = name;
			}
		}

		[Test]
		public void TestContainerIgnoresParameters()
		{
			picoContainer.RegisterComponentImplementation(typeof (Bean), typeof (NamedBean), new IParameter[]
				{
					new ConstantParameter("Dick")
				});
			picoContainer.RegisterComponentInstance("Tom");
			NamedBean bean = (NamedBean) picoContainer.GetComponentInstance(typeof (Bean));
			Assert.IsNotNull(bean);
			Assert.AreEqual("Tom", bean.Name);
		}

		[Test]
		public void TestContainerUsesStandardConstructor()
		{
			picoContainer.RegisterComponentImplementation(typeof (Bean), typeof (NamedBeanWithPossibleDefault));
			picoContainer.RegisterComponentInstance("Tom");
			NamedBeanWithPossibleDefault bean = (NamedBeanWithPossibleDefault) picoContainer.GetComponentInstance(typeof (Bean));
			Assert.IsFalse(bean.ByDefault);
		}

		[Test]
		public void TestContainerUsesOnlyStandardConstructor()
		{
			picoContainer.RegisterComponentImplementation(typeof (Bean), typeof (NoBean));
			picoContainer.RegisterComponentInstance("Tom");
			try
			{
				picoContainer.GetComponentInstance(typeof (Bean));
				Assert.Fail("Instantiation should have failed.");
			}
			catch (PicoInitializationException)
			{
			}
		}

	}
}

File: java/picocontainer/branches/picocontainer-1_2/container/src/test/org/picocontainer/defaults/SetterInjectionComponentAdapterFactoryTestCase.java
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the LICENSE.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 *****************************************************************************/
package org.picocontainer.defaults;

import org.picocontainer.Parameter;
import org.picocontainer.PicoContainer;
import org.picocontainer.PicoInitializationException;
import org.picocontainer.tck.AbstractComponentAdapterFactoryTestCase;
import org.picocontainer.tck.AbstractComponentAdapterTestCase.RecordingLifecycleStrategy;
import org.picocontainer.testmodel.NullLifecycle;
import org.picocontainer.testmodel.RecordingLifecycle;
import org.picocontainer.testmodel.RecordingLifecycle.One;

/**
 * @author J&ouml;rg Schaible</a>
 * @version $Revision$
 */
public class SetterInjectionComponentAdapterFactoryTestCase extends AbstractComponentAdapterFactoryTestCase {
    protected void setUp() throws Exception {
        picoContainer = new DefaultPicoContainer(createComponentAdapterFactory());
    }

    protected ComponentAdapterFactory createComponentAdapterFactory() {
        return new SetterInjectionComponentAdapterFactory();
    }

    public static interface Bean {
    }

    public static class NamedBean implements Bean {
        private String name;

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }
    }

    public static class NamedBeanWithPossibleDefault extends NamedBean {
        private boolean byDefault;

        public NamedBeanWithPossibleDefault() {
        }

        public NamedBeanWithPossibleDefault(String name) {
            setName(name);
            byDefault = true;
        }

        public boolean getByDefault() {
            return byDefault;
        }
    }

    public static class NoBean extends NamedBean {
        public NoBean(String name) {
            setName(name);
        }
    }

    public void testContainerUsesStandardConstructor() {
        picoContainer.registerComponentImplementation(Bean.class, NamedBeanWithPossibleDefault.class);
        picoContainer.registerComponentInstance("Tom");
        NamedBeanWithPossibleDefault bean = (NamedBeanWithPossibleDefault) picoContainer.getComponentInstance(Bean.class);
        assertFalse(bean.getByDefault());
    }

    public void testContainerUsesOnlyStandardConstructor() {
        picoContainer.registerComponentImplementation(Bean.class, NoBean.class);
        picoContainer.registerComponentInstance("Tom");
        try {
            picoContainer.getComponentInstance(Bean.class);
            fail("Instantiation should have failed.");
        } catch (PicoInitializationException e) {
        }
    }

    public void testCustomLifecycleCanBeInjected() throws NoSuchMethodException {
        RecordingLifecycleStrategy strategy = new RecordingLifecycleStrategy(new StringBuffer());
        SetterInjectionComponentAdapterFactory caf = new SetterInjectionComponentAdapterFactory(false, strategy);
        SetterInjectionComponentAdapter sica = (SetterInjectionComponentAdapter)caf.createComponentAdapter(NullLifecycle.class, NullLifecycle.class, new Parameter[0]);
        One one = new RecordingLifecycle.One(new StringBuffer());
        sica.start(one);
        sica.stop(one);        
        sica.dispose(one);
        assertEquals("<start<stop<dispose", strategy.recording());
    }    
}